src/libNode/MicroBlockPreProcessing.cpp

```diff
@@ -908,6 +1021,84 @@ void Node::UpdateBalanceForPreGeneratedAccounts() {
                         << counter);
 }
 
+void Node::StartTxnProcessingThread() {
+  LOG_MARKER();
+  auto t = [this]() -> void {
+    m_mediator.m_node->m_prePrepRunning = false;
+    if (m_mediator.ToProcessTransaction()) {
+      if (m_mediator.m_ds->m_mode == DirectoryService::Mode::IDLE) {
+        m_mediator.m_node->ProcessTransactionWhenShardBackup(
+            SHARD_MICROBLOCK_GAS_LIMIT);
+        if (!AccountStore::GetInstance().SerializeDelta()) {
+          LOG_GENERAL(WARNING, "AccountStore::SerializeDelta failed");
+        }
+      } else {
+        m_mediator.m_node->ProcessTransactionWhenShardBackup(
+            m_mediator.m_ds->m_microBlockGasLimit);
+      }
+    }
+  };
+  DetachedFunction(1, t);
+}
+
+bool Node::WaitUntilTxnProcessingDone() {
+  LOG_MARKER();
+  // wait for txn processing being ready by me (backup)
+  unique_lock<mutex> lock(m_mutexCVTxnProcFinished);
+  int timeout_time = std::max(
+      0, ((int)MICROBLOCK_TIMEOUT -
+          ((int)TX_DISTRIBUTE_TIME_IN_MS +
+           (int)((m_mediator.m_ds->m_mode == DirectoryService::Mode::IDLE)
+                     ? SHARD_ANNOUNCEMENT_DELAY_IN_MS
+                     : DS_ANNOUNCEMENT_DELAY_IN_MS)) /
+              1000 -
+          (int)CONSENSUS_OBJECT_TIMEOUT));
+  LOG_GENERAL(INFO,
+              "The overall timeout for completing txns processing will be "
+                  << timeout_time << " seconds");
+
+  while (!m_txnProcessingFinished) {
+    if (cv_TxnProcFinished.wait_for(lock, chrono::seconds(timeout_time)) ==
+        std::cv_status::timeout) {
+      // timed out
+      LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
+                "Timed out waiting for txn processing being completed. May "
+                "need to adjust timeouts.");
+      return false;
+    }
+  }
+  return true;
+}
+
+bool Node::WaitUntilCompleteMicroBlockIsReady() {
+  LOG_MARKER();
+  unique_lock<mutex> lock(m_mutexMicroBlock);
+  int timeout_time = std::max(
+      0, ((int)MICROBLOCK_TIMEOUT -
+          ((int)TX_DISTRIBUTE_TIME_IN_MS +
+           (int)((m_mediator.m_ds->m_mode == DirectoryService::Mode::IDLE)
+                     ? SHARD_ANNOUNCEMENT_DELAY_IN_MS
+                     : DS_ANNOUNCEMENT_DELAY_IN_MS)) /
```

Comment:
I believe `WaitUntilCompleteFinalBlockIsReady` is the one used by DS leader, so can probably just use `SHARD_ANNOUNCEMENT_DELAY_IN_MS`